Stores Kafka message timestamp as part of the encoded information

nachomdo opened the pull request:
This PR allows to store Kafka messages timestamp information as part of the encoded MessagePack record. 

This has been discussed here https://github.com/pinterest/secor/issues/342

HenryCaiHaiying commented on src/main/java/com/pinterest/secor/io/KeyValue.java:
```diff
@@ -50,9 +61,20 @@ public long getOffset() {
+	}
+
+	public boolean hasKafkaKey() {
+		return this.mKafkaKey.length != 0;
```
Probably should check both null and zero length

HenryCaiHaiying commented on src/main/java/com/pinterest/secor/io/impl/MessagePackSequenceFileReaderWriterFactory.java:
```diff
@@ -76,13 +77,17 @@ public KeyValue next() throws IOException {
                 MessageUnpacker unpacker = MessagePack.newDefaultUnpacker(mKey.getBytes());
                 int mapSize = unpacker.unpackMapHeader();
                 long offset = 0;
+                long timestamp = 0;
```
Should the default value of timestamp be -1, to be consistent with the KeyValue class?

HenryCaiHaiying commented on src/main/java/com/pinterest/secor/io/impl/MessagePackSequenceFileReaderWriterFactory.java:
```diff
@@ -161,5 +175,17 @@ public void write(KeyValue keyValue) throws IOException {
+            if (keyValue.hasKafkaKey())
+                fields++;
+
+            if (keyValue.hasTimestamp())
```
What if keyValue does not have kafkaKey but has timestamp, would the field index correct in that case?

HenryCaiHaiying replied:
How was this case addressed in your new fix?

nachomdo replied:
I've added a test to check the scenario you are describing (a record with timestamp but no key) 
```
KeyValue kv4 =  (new KeyValue(23235, null, new byte[]{23, 45, 40 ,10, 122}, 1496318250));
```
This method is returning the number of fields that we are encoding together, not the field index. So for a record with offset but no key and no timestamp should return 1. If we had key, offset and timestamp it should return 3 and in case either the key or the timestamp is missing it will return 2.

HenryCaiHaiying replied:
I see what you mean now, if the method only care about the number of fields, it should be fine.

nachomdo commented:
Thanks @HenryCaiHaiying for taking the time to review these changes. I have addressed your comments and added a new test case for the scenario you were describing: a record with no key but with timestamp.

HenryCaiHaiying commented:
Thanks for the contribution.
